File: core/contractsdk/java/example/source_trace/src/main/java/com/baidu/xuper/example/SourceTrace.java
package com.baidu.xuper.example;

import java.math.BigDecimal;

import com.baidu.xuper.Context;
import com.baidu.xuper.Contract;
import com.baidu.xuper.ContractMethod;
import com.baidu.xuper.Driver;
import com.baidu.xuper.Response;

/**
 * Counter
 */
public class SourceTrace implements Contract {
    final String GOODS = "GOODS_";
    final String GOODSRECORD = "GOODSRECORD_";
    final String GOODSRECORDTOP = "GOODSRECORDTOP_";
    final String CREATE = "CREATE_";

    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        String admin = new String(ctx.args().get("admin".getBytes()));
        if (admin.length() == 0) {
            return Response.error("missing admin address");
        }
        ctx.putObject("admin".getBytes(), admin.getBytes());
        return Response.ok("ok".getBytes());
    }

    private boolean isAdmin(Context ctx, String caller) {
        String admin = new String(ctx.getObject("admin".getBytes()));
        return admin.equals(caller);
    }

    private boolean isAdmin(Context ctx) {
        String caller = ctx.caller();
        return isAdmin(ctx, caller);
    }

    @ContractMethod
    public Response createGoods(Context ctx) {
        if (!isAdmin(ctx)) {
            return Response.error("only the admin can create new goods");
        }
        String id = new String(ctx.args().get("id".getBytes()));
        String desc = new String(ctx.args().get("desc".getBytes()));

        if (id.length() == 0 || desc.length() == 0) {
            return Response.error("missing id or desc");
        }
        String goodsKey = GOODS + id;
        if (ctx.getObject(goodsKey.getBytes()).length > 0) {
            return Response.error("the id  already exist, please check again");
        }
        ctx.putObject(goodsKey.getBytes(), desc.getBytes());
        String goodsRecordsKey = GOODSRECORD + id + "_0";
        String goodsRecordsTopKey = GOODSRECORDTOP + id;
        ctx.putObject(goodsRecordsKey.getBytes(), CREATE.getBytes());
        ctx.putObject(goodsRecordsTopKey.getBytes(), "0".getBytes());
        return Response.ok(id.getBytes());
    }

    @ContractMethod
    public Response updateGoods(Context ctx) {
        if (!isAdmin(ctx)) {
            return Response.error("only the admin can update goods");
        }
        String id = new String(ctx.args().get("id".getBytes()));
        String reason = new String(ctx.args().get("reason".getBytes()));
        if (id.length() == 0 || reason.length() == 0) {
            return Response.error("missing argument id or argument reason");
        }
        String goodsRecordsTopKey = GOODSRECORDTOP + id;
        BigDecimal topRecord = new BigDecimal(new String(ctx.getObject(goodsRecordsTopKey.getBytes())));
        topRecord = topRecord.add(BigDecimal.ONE);
        String goodsRecordsKey = GOODSRECORD + id + "_" + topRecord.toString();
        ctx.putObject(goodsRecordsKey.getBytes(), reason.getBytes());
        ctx.putObject(goodsRecordsTopKey.getBytes(), topRecord.toString().getBytes());

        return Response.ok(topRecord.toString().getBytes());
    }

    @ContractMethod
    public Response queryRerords(Context ctx) {
        String id = new String(ctx.args().get("id"));
        if (id.length() == 0) {
            return Response.error("missing argument id");
        }
        String goodsKey = GOODS + id;
        if (ctx.getObject(goodsKey.getBytes()).length == 0) {
            return Response.error("good with id " + id + "not found");
        }
        String goodsRecordKey = GOODSRECORD + id + "_";
        String start = goodsRecordKey;
        String end = start + "~";
        StringBuffer buf = new StringBuffer();
        buf.append("\n");

        ctx.newIterator(start.getBytes(), end.getBytes()).forEachRemaining(
                elem -> {
                    String key = elem.getKey().toString();
                    String[] goodsRecords = key.substring(GOODSRECORD.length()).split("_");
                    String goodsId = goodsRecords[0];
                    String updateRecord = goodsRecords[1];
                    String reason = new String(elem.getValue());
                    String record = "goodIds=" + goodsId + ",updateRecord=" + updateRecord + ",reason=" + reason + "\n";
                    buf.append(record.getBytes());
                }
        );
        return Response.ok(buf.toString().getBytes());
    }

    public static void main(String[] args) {
        Driver.serve(new SourceTrace());
        System.out.println(BigDecimal.ZERO.toPlainString().getBytes() == "0".getBytes());
    }
}

File: core/contractsdk/cpp/example/source_trace.cc
#include "xchain/xchain.h"

// 商品溯源合约模板
// 参数由xchain::Contract中的context提供
class SourceTrace {
public:
    /*
     * func: 初始化商品溯源合约
     * @param: admin: 哪个address具有商品管理员权限
     */
    virtual void initialize() = 0;
    /*
     * func: 创建一个新的商品
     * 说明: 仅有合约发起者为admin时才允许创建商品
     * @param: id: 商品id
     * @param: desc: 商品描述
     */
    virtual void createGoods() = 0;
    /*
     * func: 变更商品信息
     * 说明: 仅有合约发起者为admin时才允许变更商品
     * @param: id: 商品id
     * @param: reason: 变更原因
     */
    virtual void updateGoods() = 0;
    /*
     * func: 查询商品变更记录
     * @param: id: 商品id
     */
    virtual void queryRecords() = 0;
};

struct SourceTraceDemo : public SourceTrace, public xchain::Contract {
public:
    const std::string GOODS = "GOODS_";
    const std::string GOODSRECORD = "GOODSSRECORD_";
    const std::string GOODSRECORDTOP = "GOODSSRECORDTOP_";
    const std::string CREATE = "CREATE";

    void initialize() {
        xchain::Context* ctx = this->context();
        const std::string& admin = ctx->arg("admin");
        if (admin.empty()) {
            ctx->error("missing admin address");
            return;
        }

        ctx->put_object("admin", admin);
        ctx->ok("initialize success");
    }

    bool isAdmin(xchain::Context* ctx, const std::string& caller) {
        std::string admin;
        if (!ctx->get_object("admin", &admin)) {
            return false;
        }
        return (admin == caller);
    }

    void createGoods() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing initiator");
            return;
        }

        if (!isAdmin(ctx, caller)) {
            ctx->error("only the admin can create new goods");
            return;
        }

        const std::string& id = ctx->arg("id");
        if (id.empty()) {
            ctx->error("missing 'id' as goods identity");
            return;
        }

        const std::string& desc = ctx->arg("desc");
        if (desc.empty()) {
            ctx->error("missing 'desc' as goods desc");
            return;
        }

        std::string goodsKey = GOODS + id;
        std::string value;
        if (ctx->get_object(goodsKey, &value)) {
            ctx->error("the id is already exist, please check again");
            return;
        }
        ctx->put_object(goodsKey, desc);

        std::string goodsRecordsKey = GOODSRECORD + id + "_0";
        std::string goodsRecordsTopKey = GOODSRECORDTOP + id;
        ctx->put_object(goodsRecordsKey, CREATE);
        ctx->put_object(goodsRecordsTopKey, 0);
        ctx->ok(id);
    }

    void updateGoods() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing initiator");
            return;
        }

        if (!isAdmin(ctx, caller)) {
            ctx->error("only the admin can update goods");
            return;
        }

        const std::string& id = ctx->arg("id");
        if (id.empty()) {
            ctx->error("missing 'id' as goods identity");
            return;
        }

        const std::string& reason = ctx->arg("reason");
        if (reason.empty()) {
            ctx->error("missing 'reason' as update reason");
            return;
        }

        std::string goodsRecordsTopKey = GOODSRECORDTOP + id;
        std::string value;
        ctx->get_object(goodsRecordsTopKey, &value);
        if (value.length() == 0) {
            ctx->error("goods " + id + " not found");
            return;
        }
        int topRecord = 0;
        topRecord = atoi(value.c_str()) + 1;

        char topRecordVal[32];
        snprintf(topRecordVal, 32, "%d", topRecord);
        std::string goodsRecordsKey = GOODSRECORD + id + "_" + topRecordVal;

        ctx->put_object(goodsRecordsKey, reason);
        ctx->put_object(goodsRecordsTopKey, topRecordVal);
        ctx->ok(topRecordVal);
    }

    void queryRecords() {
        xchain::Context* ctx = this->context();
        const std::string& id = ctx->arg("id");
        if (id.empty()) {
            ctx->error("missing 'id' as goods identity");
            return;
        }

        std::string goodsKey = GOODS + id;
        std::string value;
        if (!ctx->get_object(goodsKey, &value)) {
            ctx->error("the id not exist, please check again");
            return;
        }

        std::string goodsRecordsKey = GOODSRECORD + id + "_";
        std::unique_ptr<xchain::Iterator> iter =
            ctx->new_iterator(goodsRecordsKey, goodsRecordsKey + "~");

        std::string result = "\n";
        while (iter->next()) {
            std::pair<std::string, std::string> res;
            iter->get(&res);
            if (res.first.length() > goodsRecordsKey.length()) {
                std::string goodsRecord =
                    res.first.substr(GOODSRECORD.length());
                std::string::size_type pos = goodsRecord.find("_");
                std::string goodsId = goodsRecord.substr(0, pos);
                std::string updateRecord =
                    goodsRecord.substr(pos + 1, goodsRecord.length());
                std::string reason = res.second;

                result += "goodsId=" + goodsId +
                          ",updateRecord=" + updateRecord +
                          ",reason=" + reason + '\n';
            }
        }
        ctx->ok(result);
    }
};

DEFINE_METHOD(SourceTraceDemo, initialize) { self.initialize(); }

DEFINE_METHOD(SourceTraceDemo, createGoods) { self.createGoods(); }

DEFINE_METHOD(SourceTraceDemo, updateGoods) { self.updateGoods(); }

DEFINE_METHOD(SourceTraceDemo, queryRecords) { self.queryRecords(); }
